examples/manifold_gradients.dx

```diff
@@ -0,0 +1,342 @@
+' ## Gradients on manifolds
+An example of how to generalize the Dex `linearize` and `grad` functions to take
+derivatives and gradients of functions on manifolds.
+
+' ### Tangent space
+The difference between differentiating on $\mathbb{R}^n$ and on general
+manifolds is that in the general case, the manifold tangent space type may be
+different to the manifold representation type. The tangent space at a single
+point will always be isomorphic to $\mathbb{R}^n$ for some $n$, but here we
+allow any vector space type `b` for flexibility.
+
+' The proper definition of a tangent space is quite abstract, so no further details will be provided here. Wikipedia has quite a good
+[article](https://en.wikipedia.org/wiki/Tangent_space) for more information.
+
+' Practically, the two components required for defining the differentiable
+structure on the manifold are the exponential map (a mapping from the tangent
+space at p to the manifold, such that zero is mapped to $p$, and straight lines
+are mapped to geodesics), and its inverse. Note that "exponential map" is a
+somewhat overloaded term, here we are talking about it in a Riemannian geometry
+sense.
+
+' In the `TangentSpace` interface, `a` is the type which can represent points
+on the manifold. `b` is the type which can represent points in a tangent space
+to the manifold.  Ideally `b` would be an associated type of `a` so that the
+compiler is able to infer the tangent space type from the manifold type, but
+this feature is not yet available in Dex.
+
+interface [VSpace b] TangentSpace a b
+  -- Given a point p on the manifold M and a point v in its tangent space TpM,
+  -- this returns another point q on the manifold.
+  -- Must satisfy tangentExponentialMap p zero == p
+  -- (tangentExponentialMap p) will be linearized around zero, so it is
+  -- important that any operations used have well-defined derivatives for inputs
+  -- p, zero.
+  tangentExponentialMap: a -> b -> a
+  -- (tangentLogarithmMap p) is the pointwise inverse to
+  -- (tangentExponentialMap p).
+  -- Must satisfy tangentLogarithmMap p p == zero
+  -- (tangentLogarithmMap p) will be linearized around p, so it is important
+  -- that any operations used have well-defined derivatives for inputs p, p.
+  tangentLogarithmMap: a -> a -> b
+
+
+'### Differentiation functions
+
+'We can then define a version of the Dex `linearize` function which works on
+manifolds.  To see intuitively why this works, consider the following diagram.
+We would like to obtain a linearized version of the function "across the top" of
+the diagram, and we can form this by linearizing the function formed from the
+bottom three sides of the square.
+
+:html
+   "<img
+     src=\"https://user-images.githubusercontent.com/33986545/122416078-5a23ef00-cf80-11eb-83d5-d0d5986e8d20.png\"
+     width=\"100%\">"
```

Comment:
Where's that image from? Did you draw it? Also, isn't it possible to just put the `<img>` tag in a literate markdown block?

Reply:
Yep, hand drawn, I didn't fancy working out the LaTeX diagram... 

I could've used the markdown `img` tag but I couldn't find a way to make the width auto-scale, whereas the `100%` tag seems to work in the html.